### PR TITLE
feature: schema resolver

### LINKED ISSUE #1855: Using custom AJV breaks resolving references
## 🐛 Bug Report

When using a custom AJV instance as described in the documentation (https://www.fastify.io/docs/latest/Validation-and-Serialization/#schema-compiler) one cannot provide a schema with references as a response schema.

## To Reproduce

```js
'use strict';

const AJV = require('ajv');
const ajv = new AJV();

ajv.addSchema({
  $id: 'urn:schema:foo',
  definitions: {
    foo: { type: 'string' }
  },

  type: 'object',
  properties: {
    foo: { $ref: '#/definitions/foo' }
  }
});

ajv.addSchema({
  $id: 'urn:schema:response',
  type: 'object',
  properties: {
    foo: { $ref: 'urn:schema:foo#/definitions/foo' }
  }
});

const fastify = require('fastify')();
fastify.setSchemaCompiler(schema => ajv.compile(schema));

fastify.route({
  path: '/',
  method: 'GET',
  schema: {
    response: {
      '2xx': ajv.getSchema('urn:schema:response').schema
    }
  },
  handler(req, reply) {
    reply.send({ foo: 'bar' });
  }
});

fastify.inject({ url: '/' }, (err, response) => {
  if (err) {
    throw err;
  }
  console.log(response.body);
});
```

## Expected behavior

The route should register and resolve references correctly.

Paste the results here:

```js
/private/tmp/17/fastify-custom-ajv/index.js:44
    throw err;
    ^
FastifyError [FST_ERR_SCH_BUILD]: FST_ERR_SCH_BUILD: Failed building the schema for GET: /, due error Cannot read property '$ref' of undefined
    at Object.afterRouteAdded (/private/tmp/17/fastify-custom-ajv/node_modules/.registry.npmjs.org/fastify/2.8.0/node_modules/fastify/lib/route.js:224:37)
    at after (/private/tmp/17/fastify-custom-ajv/node_modules/.registry.npmjs.org/fastify/2.8.0/node_modules/fastify/lib/route.js:165:25)
    at Object._encapsulateThreeParam (/private/tmp/17/fastify-custom-ajv/node_modules/.registry.npmjs.org/avvio/6.2.2/node_modules/avvio/boot.js:417:7)
    at Boot.callWithCbOrNextTick (/private/tmp/17/fastify-custom-ajv/node_modules/.registry.npmjs.org/avvio/6.2.2/node_modules/avvio/boot.js:339:5)
    at Boot._after (/private/tmp/17/fastify-custom-ajv/node_modules/.registry.npmjs.org/avvio/6.2.2/node_modules/avvio/boot.js:230:26)
    at Plugin.exec (/private/tmp/17/fastify-custom-ajv/node_modules/.registry.npmjs.org/avvio/6.2.2/node_modules/avvio/plugin.js:89:17)
    at Boot.loadPlugin (/private/tmp/17/fastify-custom-ajv/node_modules/.registry.npmjs.org/avvio/6.2.2/node_modules/avvio/plugin.js:175:10)
    at Task.release (/private/tmp/17/fastify-custom-ajv/node_modules/.registry.npmjs.org/fastq/1.6.0/node_modules/fastq/queue.js:127:16)
    at worked (/private/tmp/17/fastify-custom-ajv/node_modules/.registry.npmjs.org/fastq/1.6.0/node_modules/fastq/queue.js:169:10)
    at toLoad.finish (/private/tmp/17/fastify-custom-ajv/node_modules/.registry.npmjs.org/avvio/6.2.2/node_modules/avvio/plugin.js:178:7)
```

## Your Environment

- *node version*: 10.16.3
- *fastify version*: 2.8.0
- *os*: Mac

### PR DESCRIPTION
**Pull request by Eomm**
Fixes #1855 

This is the quick&dirty version to check the doability.
Missing:
- [x] clean code
- [x] test of `setSchemaResolver`
- [x] test resolver for validation schema (query, params, body, headers)
- [ ] update docs

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by jsumners** [comment]
You’re amazing. Thank you very much for taking the time to work on this. It’s looking good so far.

File: lib/route.js [hunk 1]
```diff
@@ -218,8 +219,10 @@ function buildRouting (options) {
             this.setSchemaCompiler(buildSchemaCompiler(externalSchemas, schemaCache))
           }
 
-          buildSchema(context, opts.schemaCompiler || this[kSchemaCompiler], this[kSchemas])
+          buildSchema(context, opts.schemaCompiler || this[kSchemaCompiler], this[kSchemas], this[kSchemaResolver])
         } catch (error) {
+          console.log({ error })
```
Comment (mcollina) [inline lib/route.js]: This should likely be removed.

**Comment by Eomm** [comment]
> You’re amazing. Thank you very much for taking the time to work on this. It’s looking good so far.

It's a pleasure to help 💪
I need to add some other use cases and docs.. I have this in mind:


image

**Comment by Eomm** [comment]
⚠ The schema resolver can't work without a client's schema compiler.

I have tried to use the default schema compiler and, in the schema resolver run this:

```
fastify.setSchemaResolver((ref) => {
  fastify.schemaCompiler.addSchema(myschema)
})
```

but there is a different order in loading:

- the fastify server is starting
- the first route starts loading
   - the default schema compiler (aka `compile` function) is built with the first route in the fastify instance and it is assigned
   - the route start building the validation and serialization function
   - the input route `schemas.js` resolve the `shared schema` link, without involving the default compiler
   - the schema added to fastify with `addSchema` and the schema resolved are provided to `fast-json-stringify` to build the response serialization function
   - the schema is provided to the `compile` function for body, params, query, headers**XXX**
- the other instance routes start loading
   - the instance compiler is used
   - as above the `schema.js` resolves the shared schema
   - etc... the same


**XXX** : the compile function is already generated, so we cannot modify it, so it is too late to run `fastify.schemaCompiler.addSchema(myschema)`

File: docs/Validation-and-Serialization.md [hunk 2]
```diff
@@ -301,6 +301,68 @@ In that case the function returned by `schemaCompiler` returns an object like:
 * `error`: filled with an instance of `Error` or a string that describes the validation error
 * `value`: the coerced value that passed the validation
 
+<a name="schema-resolver"></a>
+#### Schema Resolver
+
+The `schemaResolver` is a function that works togheter with the `schemaCompiler`: you can't use it
```
Comment (jsumners) [inline docs/Validation-and-Serialization.md]:
```suggestion
The `schemaResolver` is a function that works together with the `schemaCompiler`: you can't use it
```

Comment (mcollina) [reply docs/Validation-and-Serialization.md]: @Eomm can you apply this so that we can land it and release in 2.9.0?

**Pull request merged**